include/ugfx.h

```diff
@@ -0,0 +1,601 @@
+#define ugfx_force_matrix(slot, matrix) ugfx_load_matrix(slot, matrix, UGFX_MTX_FORCE)
+
+#define ugfx_load_viewport(slot, viewport) (\
+    __ugfx_opcode(UGFX_OP_LOAD_VIEWPORT) | \
+    __ugfx_mask_shift((slot), 0xF, 28) | \
+    __ugfx_mask_shift((uintptr_t)(viewport), 0x1FFFFFF, 0))
+
+#define ugfx_load_light(slot, light, index) (\
+    __ugfx_opcode(UGFX_OP_LOAD_LIGHT) | \
+    __ugfx_mask_shift((index), 0x7, 35) | \
+    __ugfx_mask_shift((slot), 0xF, 28) | \
+    __ugfx_mask_shift((uintptr_t)(light), 0x1FFFFFF, 0))
+
+#define ugfx_link_commands(slot, commands, length, flags) (\
+    __ugfx_opcode(UGFX_OP_LINK_COMMANDS) | \
+    __ugfx_mask_shift((length), 0x1FFFFF, 35) | \
+    __ugfx_mask_shift((flags), 0x7, 32) | \
+    __ugfx_mask_shift((slot), 0xF, 28) | \
+    __ugfx_mask_shift((uintptr_t)(commands), 0x1FFFFFF, 0))
+
+#define ugfx_push_commands(slot, commands, length) ugfx_link_commands(slot, commands, length, UGFX_LINK_PUSH)
+#define ugfx_load_commands(slot, commands, length) ugfx_link_commands(slot, commands, length, UGFX_LINK_LOAD)
+
+#define ugfx_set_cull_mode(mode) (\
+    __ugfx_opcode(UGFX_OP_SET_CULL_MODE) | \
+    __ugfx_mask_shift((mode), 0x3, 32))
+
+#define ugfx_set_geometry_mode(mode) (\
+    __ugfx_opcode(UGFX_OP_SET_GEOMETRY_MODE) | \
+    __ugfx_mask_shift((mode), 0xFF, 32))
+
+#define ugfx_set_persp_norm(scale) (\
+    __ugfx_opcode(UGFX_OP_SET_PERSP_NORM) | \
+    __ugfx_mask_shift((scale), 0xFFFF, 32))
+
+#define ugfx_set_texture_settings(scale_s, scale_t, level, tile) (\
+    __ugfx_opcode(UGFX_OP_SET_TEXTURE_SETTINGS) | \
+    __ugfx_mask_shift((level), 0x7, 35) | \
+    __ugfx_mask_shift((tile), 0x7, 32) | \
+    __ugfx_mask_shift((scale_s), 0xFFFF, 16) | \
+    __ugfx_mask_shift((scale_t), 0xFFFF, 0))
+
+#define ugfx_set_num_lights(num) (\
+    __ugfx_opcode(UGFX_OP_SET_NUM_LIGHTS) | \
+    __ugfx_mask_shift((num), 0x7, 35))
+
+#define ugfx_set_clip_ratio(ratio) (\
+    __ugfx_opcode(UGFX_OP_SET_CLIP_RATIO) | \
+    __ugfx_mask_shift((ratio), 0xFFFF, 32))
+
+#define ugfx_set_address_slot(slot, address) (\
+    __ugfx_opcode(UGFX_OP_SET_ADDRESS_SLOT) | \
+    __ugfx_mask_shift((slot), 0xF, 28) | \
+    __ugfx_mask_shift((uintptr_t)(address), 0x1FFFFFF, 0))
+
+#define ugfx_draw_triangle(v0, v1, v2) ( \
+    __ugfx_opcode(UGFX_OP_DRAW_TRIANGLE) | \
+    __ugfx_mask_shift((v0), 0x3F, 49) | \
+    __ugfx_mask_shift((v1), 0x3F, 43) | \
+    __ugfx_mask_shift((v2), 0x3F, 37))
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct ugfx_vertex_t {
+    int16_t x;
+    int16_t y;
+    int16_t z;
+    int16_t padding;
+    int16_t s;
+    int16_t t;
+
+    union {
+        uint32_t rgba;
+
+        struct {
+          uint8_t r;
+          uint8_t g;
+          uint8_t b;
+          uint8_t a;
+        } color;
+
+        struct {
+          int8_t x;
+          int8_t y;
+          int8_t z;
+          uint8_t a;
+        } normal;
+    } attr;
+} __attribute__((aligned(16))) ugfx_vertex_t;
```

Comment:
I generally like this idea. I was thinking about going even further and separating each vertex component into their own buffers (so one could choose the vertex format on the fly without having to rearrange the vertex buffer). But that would make efficient DMA transfers very difficult, if not impossible. It's only possible to specify stride and line width on the source, but not on the target, correct? This is just an idea, though. I'm not sure if this additional effort is even worth it, or would serve any valid use cases.

Reply:
That could be an option as well, yes, though you would need multiple DMA transfers to fetch a vertex, which might get a bit in the way. I would be happy to have a single vertex buffer with a descriptor, similar to how OpenGL handles the vertex buffer objects. It'd be a more static version of course, for performance, so I don't think we need to allow any kind of arrangement of data. Just having a bitmask to turn on/off attributes (like optional fields of a single structure) I think it would be enough.